File: gcs/m3gcs/m3psu.py
from .packets import register_packet, register_command
import struct


def msg_id(x):
    return x << 5


CAN_ID_M3PSU = 2
CAN_MSG_ID_M3PSU_BATT_VOLTAGES = CAN_ID_M3PSU | msg_id(56)
CAN_MSG_ID_M3PSU_TOGGLE_PYROS = CAN_ID_M3PSU | msg_id(16)
CAN_MSG_ID_M3PSU_CHANNEL_STATUS_12 = CAN_ID_M3PSU | msg_id(49)
CAN_MSG_ID_M3PSU_CHANNEL_STATUS_34 = CAN_ID_M3PSU | msg_id(50)
CAN_MSG_ID_M3PSU_CHANNEL_STATUS_56 = CAN_ID_M3PSU | msg_id(51)
CAN_MSG_ID_M3PSU_CHANNEL_STATUS_78 = CAN_ID_M3PSU | msg_id(52)
CAN_MSG_ID_M3PSU_CHANNEL_STATUS_910 = CAN_ID_M3PSU | msg_id(53)
CAN_MSG_ID_M3PSU_CHANNEL_STATUS_1112= CAN_ID_M3PSU | msg_id(54)
CAN_MSG_ID_M3PSU_TOGGLE_CHANNEL = CAN_ID_M3PSU | msg_id(17)
CAN_MSG_ID_M3PSU_PYRO_STATUS = CAN_ID_M3PSU | msg_id(48)
CAN_MSG_ID_M3PSU_CHARGER_STATUS = CAN_ID_M3PSU | msg_id(55)
CAN_MSG_ID_M3PSU_TOGGLE_CHARGER = CAN_ID_M3PSU | msg_id(18)
CAN_MSG_ID_M3PSU_TOGGLE_LOWPOWER = CAN_ID_M3PSU | msg_id(19)
CAN_MSG_ID_M3PSU_TOGGLE_BATTLESHORT = CAN_ID_M3PSU | msg_id(20)
CAN_MSG_ID_M3PSU_CAPACITY = CAN_ID_M3PSU | msg_id(57)


@register_packet("m3psu", CAN_MSG_ID_M3PSU_BATT_VOLTAGES,
    "Battery Voltages")
def batt_volts(data):
    # 3 bytes
    cell1, cell2, batt = struct.unpack("HHH", bytes(data[:6]))
    cell1 *= 0.01
    cell2 *= 0.01
    batt *= 0.01

    return "{: 4.2f}V, {: 4.2f}V, Batt: {: 4.2f}V".format(cell1, cell2, batt)


@register_packet("m3psu", CAN_MSG_ID_M3PSU_TOGGLE_PYROS, "Toggle Pyros")
def toggle_pyros(data):
    if data[0] == 0:
        return "Disable pyros"
    elif data[0] == 1:
        return "Enable pyros"
    else:
        return "Invalid packet"


@register_packet("m3psu", CAN_MSG_ID_M3PSU_CHANNEL_STATUS_12,
    "Channel 1,2 status")
@register_packet("m3psu", CAN_MSG_ID_M3PSU_CHANNEL_STATUS_34,
    "Channel 3,4 status")
@register_packet("m3psu", CAN_MSG_ID_M3PSU_CHANNEL_STATUS_56,
    "Channel 5,6 status")
@register_packet("m3psu", CAN_MSG_ID_M3PSU_CHANNEL_STATUS_78,
    "Channel 7,8 status")
@register_packet("m3psu", CAN_MSG_ID_M3PSU_CHANNEL_STATUS_910,
    "Channel 9,10 status")
@register_packet("m3psu", CAN_MSG_ID_M3PSU_CHANNEL_STATUS_1112,
    "Channel 11,12 status")
def channel_status(data):
    # 8 bytes: voltage/0.03V, current/0.003A, power/0.02W, blank,
    # voltage/0.03V, current/0.003A, power/0.02W, blank
    voltage, current, power = struct.unpack("BBBx", bytes(data[:4]))
    string = "{: 5.3f}V {: 6.3f}A {: 5.2f}W".format(voltage * 0.03,
        current * 0.003, power * 0.02)
    voltage, current, power = struct.unpack("BBBx", bytes(data[4:]))
    string += ", {: 5.3f}V {: 6.3f}A {: 5.2f}W".format(voltage * 0.03,
        current * 0.003, power * 0.02)
    return string


@register_packet("m3psu", CAN_MSG_ID_M3PSU_TOGGLE_CHANNEL,
    "Toggle Channel")
def toggle_channel(data):
    if data[0] == 0:
        return "Disable channel {: 2d}".format(data[1])
    elif data[1] == 1:
        return "Enable channel {: 2d}".format(data[1])
    else:
        return "Invalid packet"


@register_packet("m3psu", CAN_MSG_ID_M3PSU_PYRO_STATUS, "Pyro Status")
def pyro_status(data):
    # 7 bytes: 16bit voltage (mV), 16bit current (uA), 16bit power (0.1mW)
    # 1bit pyro enable line measurement
    voltage, current, power, pyro = struct.unpack("HHHB", bytes(data[:7]))
    string = "{: 5.3f}V {: 6.3f}A {: 7.3f}W".format(voltage/1000.0,
        current/1000000.0, power/100000.0)
    if pyro == 1:
        string += ", pyro enabled"
    elif pyro == 0:
        string += ", pyro disabled"
    else:
        string += ", invalid packet!"
    return string

@register_packet("m3psu", CAN_MSG_ID_M3PSU_CHARGER_STATUS, "Charger Status")
def charger_status(data):
    # 5 bytes. First two are charge current in mA, 3rd is status bits, 4-5th are temperature in cK
    current, state, tempcK = struct.unpack("=hBH", bytes(data[:5]))
    charger_enabled = bool(state & 1)
    is_charging = bool(state & 2)
    charge_inhibit = bool(state & 4)
    battleshort = bool(state & 32)
    voltage_mode = (state >> 3) & 0x3;
    tempC = (tempcK/10) - 273.2

    string = "{: 4d}mA, {: 3.1f}degC".format(current, tempC)
    if charger_enabled:
        string += ", charger enabled"
    else:
        string += ", charger disabled"
    if is_charging:
        string += ", charging"
    if charge_inhibit:
        string += ", inhibited"
    if battleshort:
        string += ", WAR MODE"
    else:
        string += ", peace mode"
    string += ", {} mode".format(["PCV", "LV", "MV", "HV", "INVAL"][voltage_mode])
    return string

@register_packet("m3psu", CAN_MSG_ID_M3PSU_TOGGLE_CHARGER, "Toggle Charger")
def toggle_charger(data):
    if data[0] == 0:
        return "Disable charger"
    elif data[0] == 1:
        return "Enable charger"
    else:
        return "Invalid packet"

@register_packet("m3psu", CAN_MSG_ID_M3PSU_CAPACITY, "Capacity")
def capacity(data):
    mins, = struct.unpack("h", bytes(data[:2]))
    if mins == -1:
        mins = "Inf"
    percent = data[2]
    return "Capacity: {}%, Time left: {}".format(percent, mins)


@register_command("m3psu", "Pyro supply", ("Off", "On"))
def toggle_pyros_cmd(data):
    data = [{"Off":0, "On":1}[data]]
    return CAN_MSG_ID_M3PSU_TOGGLE_PYROS, data

@register_command("m3psu", "Charger", ("Off", "On"))
def toggle_charger_cmd(data):
    data = [{"Off":0, "On":1}[data]]
    return CAN_MSG_ID_M3PSU_TOGGLE_CHARGER, data

@register_command("m3psu", "Lowpower", ("Off", "On"))
def toggle_lowpower_cmd(data):
    data = [{"Off":0, "On":1}[data]]
    return CAN_MSG_ID_M3PSU_TOGGLE_LOWPOWER, data

@register_command("m3psu", "Battleshort", ("Peace", "War"))
def toggle_battleshort(data):
    data = [{"Peace":0, "War":1}[data]]
    return CAN_MSG_ID_M3PSU_TOGGLE_BATTLESHORT, data

@register_command("m3psu", "5V IMU", ("1 Off", "1 On"))
@register_command("m3psu", "5V AUX 2", ("2 Off", "2 On"))
@register_command("m3psu", "3V3 FC", ("3 Off", "3 On"))
@register_command("m3psu", "3V3 IMU", ("4 Off", "4 On"))
@register_command("m3psu", "5V Radio", ("5 Off", "5 On"))
@register_command("m3psu", "5V AUX 1", ("6 Off", "6 On"))
@register_command("m3psu", "3V3 Pyro", ("7 Off", "7 On"))
@register_command("m3psu", "3V3 Radio", ("8 Off", "8 On"))
@register_command("m3psu", "5V Cameras", ("9 Off", "9 On"))
@register_command("m3psu", "3V3 AUX 1", ("10 Off", "10 On"))
@register_command("m3psu", "3V3 DL", ("11 Off", "11 On"))
@register_command("m3psu", "5V CAN", ("12 Off", "12 On"))
def toggle_channel_cmd(data):
    [channel, operation] = data.split(" ")
    data = [{"Off":0, "On":1}[operation], int(channel)-1]
    return CAN_MSG_ID_M3PSU_TOGGLE_CHANNEL, data

File: gcs/m3gcs/static/parsers/m3psu.js
var M3PSU = function(gcs){
    var CAN_ID_M3PSU = 2;
    var CAN_MSG_ID_M3PSU_STATUS = CAN_ID_M3PSU | msg_id(0);
    var CAN_MSG_ID_M3PSU_TOGGLE_PYROS = CAN_ID_M3PSU | msg_id(16);
    var CAN_MSG_ID_M3PSU_TOGGLE_CHANNEL = CAN_ID_M3PSU | msg_id(17);
    var CAN_MSG_ID_M3PSU_TOGGLE_CHARGER = CAN_ID_M3PSU | msg_id(18);
    var CAN_MSG_ID_M3PSU_PYRO_STATUS = CAN_ID_M3PSU | msg_id(48);
    var CAN_MSG_ID_M3PSU_CHANNEL_STATUS_12 = CAN_ID_M3PSU | msg_id(49);
    var CAN_MSG_ID_M3PSU_CHANNEL_STATUS_34 = CAN_ID_M3PSU | msg_id(50);
    var CAN_MSG_ID_M3PSU_CHANNEL_STATUS_56 = CAN_ID_M3PSU | msg_id(51);
    var CAN_MSG_ID_M3PSU_CHANNEL_STATUS_78 = CAN_ID_M3PSU | msg_id(52);
    var CAN_MSG_ID_M3PSU_CHANNEL_STATUS_910 = CAN_ID_M3PSU | msg_id(53);
    var CAN_MSG_ID_M3PSU_CHANNEL_STATUS_1112= CAN_ID_M3PSU | msg_id(54);
    var CAN_MSG_ID_M3PSU_CHARGER_STATUS = CAN_ID_M3PSU | msg_id(55);
    var CAN_MSG_ID_M3PSU_BATT_VOLTAGES = CAN_ID_M3PSU | msg_id(56);
    var CAN_MSG_ID_M3PSU_CAPACITY = CAN_ID_M3PSU | msg_id(57);
    var CAN_MSG_ID_M3PSU_AWAKE_TIME = CAN_ID_M3PSU | msg_id(58);

    var components = {
        0: "DCDC1",
        1: "DCDC2",
        2: "DCDC3",
        3: "DCDC4",
        4: "DCDC5",
        5: "DCDC6",
        6: "Charger",
        7: "Pyro Monitor",
    };

    var dcdc_errors = {
        0: "No Error",
        1: "Init",
        2: "Ch1 Alert",
        3: "Ch2 Alert",
        4: "Comms",
    };

    var charger_errors = {
        0: "No Error",
        1: "Init",
        2: "Read",
    };

    var pyro_mon_errors = {
        0: "No Error",
        1: "Init",
        2: "Comms",
    };

    // State
    this.status = {overall: 0, components: {}};
    for(var i in components){
        var c = components[i];
        this.status.components[c] = {state: 0, reason: "Unknown"};
    }

    this.channels = [];
    for(var i=0; i<12; i++){
        this.channels[i] = {v:0, i:0, p:0};
    }

    this.charger = {
        enabled: false,
        charging: false,
        inhibit: false,
        battleshort: false,
        acfet: false,
        voltage_mode: "low",
        temperature: 0,
        current: 0,
    };

    this.runtime = 0;
    this.percent = 0;

    this.cells = [0, 0];
    this.batt_voltage = 0;

    this.pyro = {
        v:0, i:0, p:0,
        enabled: false,
    };

    this.awake_time = 0;
    this.power_mode = "low";

    var _this = this;

    //TODO handle registering commands?
    gcs.registerPacket(CAN_MSG_ID_M3PSU_TOGGLE_CHANNEL, function(data){} );

    gcs.registerPacket(CAN_MSG_ID_M3PSU_TOGGLE_PYROS, function(data){} );

    gcs.registerPacket(CAN_MSG_ID_M3PSU_TOGGLE_CHARGER, function(data){} );

    // Packet parsers
    gcs.registerPacket(CAN_MSG_ID_M3PSU_STATUS, function(data){
        var [overall, comp, comp_state, comp_error] = gcs.struct.Unpack("BBBB", data);
        var c = components[comp];
        var c_status = _this.status.components[c];
        c_status.state = comp_state;
        if(comp >= 0 && comp <= 5){
            c_status.reason = dcdc_errors[comp_error];
        }else if(comp == 6){
            c_status.reason = charger_errors[comp_error];
        }else if(comp == 7){
            c_status.reason = pyro_mon_errors[comp_error];
        }

        _this.status.overall = ["OK", "INIT", "ERROR"][overall];
    });

    gcs.registerPacket(CAN_MSG_ID_M3PSU_PYRO_STATUS, function(data){
        var [v, i, p, state] = gcs.struct.Unpack("<HHHb", data);
        _this.pyro.v = v * 0.001;
        _this.pyro.i = i * 0.001;;
        _this.pyro.p = p * 0.001;;
        _this.pyro.enabled = bool(state & 1);
    });

    var handleChannels = function(c, data){
        var c1 = _this.channels[c];
        var c2 = _this.channels[c+1];

        c1.v = data[0] * 0.03;
        c1.i = data[1] * 0.0003;
        c1.p = data[2] * 0.002;

        c2.v = data[4] * 0.03;
        c2.i = data[5] * 0.0003;
        c2.p = data[6] * 0.002;
    };
    gcs.registerPacket(CAN_MSG_ID_M3PSU_CHANNEL_STATUS_12, function(data){
        handleChannels(0, data);
    });
    gcs.registerPacket(CAN_MSG_ID_M3PSU_CHANNEL_STATUS_34, function(data){
        handleChannels(2, data);
    });
    gcs.registerPacket(CAN_MSG_ID_M3PSU_CHANNEL_STATUS_56, function(data){
        handleChannels(4, data);
    });
    gcs.registerPacket(CAN_MSG_ID_M3PSU_CHANNEL_STATUS_78, function(data){
        handleChannels(6, data);
    });
    gcs.registerPacket(CAN_MSG_ID_M3PSU_CHANNEL_STATUS_910, function(data){
        handleChannels(8, data);
    });
    gcs.registerPacket(CAN_MSG_ID_M3PSU_CHANNEL_STATUS_1112, function(data){
        handleChannels(10, data);
    });

    gcs.registerPacket(CAN_MSG_ID_M3PSU_CHARGER_STATUS, function(data){
        var [current, state, tempcK]= gcs.struct.Unpack("<hBH", data);

        var vmode = (state >> 3) & 0x3;

        var c = _this.charger;
        c.current = current;
        c.enabled = bool(state & 1);
        c.charging = bool(state & 2);
        c.inhibit = bool(state & 4);
        c.battleshort = bool(state & 32);
        c.acfet = bool(state & 64);
        c.voltage_mode = ["Pre-charge", "Low", "Med", "High", "INVAL"][vmode];
        c.temperature = (tempcK/10) - 273.2;
    });

    gcs.registerPacket(CAN_MSG_ID_M3PSU_BATT_VOLTAGES, function(data){
        var [v1, v2, vb] = gcs.struct.Unpack("<HHH", data);
        _this.cells[0] = v1 * 0.01;
        _this.cells[1] = v2 * 0.01;
        _this.batt_voltage = vb * 0.01;
    });

    gcs.registerPacket(CAN_MSG_ID_M3PSU_CAPACITY, function(data){
        var [mins, percent] = gcs.struct.Unpack("<HB", data);

        if(mins == 65535){ // Charging, so runtime is infinite
            mins = Infinity;
        }

        _this.percent = percent;
        _this.runtime = mins;
    });

    gcs.registerPacket(CAN_MSG_ID_M3PSU_AWAKE_TIME, function(data){
        var [secs, stats] = gcs.struct.Unpack("<HB", data);

        _this.awake_time = secs;
        _this.power_mode = (status & 1) ? "low" : "high";
    });
}
